lib/charms/opensearch/v0/opensearch_distro.py

```diff
@@ -126,13 +125,84 @@ def start(self):
         """Start the opensearch service."""
         pass
 
-    @abstractmethod
     def restart(self):
         """Restart the opensearch service."""
-        pass
+        if self.is_started():
+            self.stop()
+
+        self.start()
 
-    @abstractmethod
     def stop(self):
+        """Exclude the allocation of this node."""
+        try:
+            self.add_allocation_exclusions(self._charm.unit_name)
+
+            response = self.request("GET", "/_cluster/health?wait_for_status=green&timeout=60s")
+            unassigned_shards = response.get("unassigned_shards", 0)
+            if unassigned_shards > 0:
+                self._charm.on_unassigned_shards(unassigned_shards)
+
+            # stop the opensearch service
+            self._stop_service()
+
+            if self._charm.alternative_host:
+                # remove the exclusion back
+                self.remove_allocation_exclusions(
+                    self._charm.unit_name, self._charm.alternative_host
+                )
+            else:
+                # no node online, store in the app databag to exclude at a future start
+                self._charm.append_allocation_exclusion_to_remove(self._charm.unit_name)
+        except Exception as e:
```

Comment:
Can a catch all here could lead to many situations where opensearch becomes bricked? Say with wrong exclusions or with incorrect data on the databag. If so, I would add a blocked status for the charm where this is being handled.

Reply:
Thanks Raul, good catch!